src/jump_moi_overloads.jl

```diff
@@ -1,6 +1,210 @@
     JuMP.check_belongs_to_model(func, model)
-    return MOI.set(model, ForwardInObjective(), JuMP.moi_function(func))
+    return MOI.set(model, attr, JuMP.moi_function(func))
+end
+
+"""
+    abstract type AbstractLazyScalarFunction <: MOI.AbstractScalarFunction end
+
+Subtype of `MOI.AbstractScalarFunction` that is not a standard MOI scalar
+function but can be converted to one using [`standard_form`](@ref).
+
+The function can also be inspected lazily using `JuMP.coefficient` or
+[`quad_sym_half`](@ref).
+"""
+abstract type AbstractLazyScalarFunction <: MOI.AbstractScalarFunction end
+
+"""
+    standard_form(func::AbstractLazyScalarFunction)
+
+Converts `func` to a standard MOI scalar function.
+
+    standard_form(func::MOItoJuMP)
+
+Converts `func` to a standard JuMP scalar function.
+"""
+function standard_form end
+
+"""
+    quad_sym_half(func, vi1::MOI.VariableIndex, vi2::MOI.VariableIndex)
+
+Return `Q[i,j] = Q[j,i]` where the quadratic terms of `func` is represented
+by `x' Q x / 2` for a symmetric matrix `Q` where `x[i] = vi1` and `x[j] = vi2`.
+Note that while this is equal to `JuMP.coefficient(func, vi1, vi2)` if `vi1 != vi2`,
+in the case `vi1 == vi2`, it is rather equal to
+`2JuMP.coefficient(func, vi1, vi2)`.
+"""
+function quad_sym_half end
+
+standard_form(func::Union{MOI.SingleVariable,MOI.ScalarAffineFunction,MOI.ScalarQuadraticFunction}) = func
+function Base.isapprox(func1::AbstractLazyScalarFunction, func2::MOI.AbstractScalarFunction; kws...)
+    return isapprox(standard_form(func1), standard_form(func2); kws...)
+end
+
+# In the future, we could replace by https://github.com/jump-dev/MathOptInterface.jl/pull/1238
+"""
+    struct VectorScalarAffineFunction{T, VT} <: MOI.AbstractScalarFunction
+        terms::VT
+        constant::T
+    end
+
+Represents the function `x ⋅ terms + constant`
+as an `MOI.AbstractScalarFunction` where `x[i] = MOI.VariableIndex(i)`.
+Use [`standard_form`](@ref) to convert it to a `MOI.ScalarAffineFunction{T}`.
+"""
+struct VectorScalarAffineFunction{T, VT} <: MOI.AbstractScalarFunction
+    terms::VT
+    constant::T
+end
+function JuMP.coefficient(func::VectorScalarAffineFunction, vi::MOI.VariableIndex)
+    return func.terms[vi.value]
+end
+function Base.convert(::Type{MOI.ScalarAffineFunction{T}}, func::VectorScalarAffineFunction) where {T}
+    n = length(func.terms)
+    return MOI.ScalarAffineFunction{T}(
+        # TODO we should do better if the vector is a `SparseVector`, I think
+        #      I have some code working for both vector types in Polyhedra.jl
+        MOI.ScalarAffineTerm{T}[
+            MOI.ScalarAffineTerm{T}(func.terms[i], VI(i))
+            for i in eachindex(func.terms) if !iszero(func.terms[i])
+        ],
+        func.constant,
+    )
+end
+function standard_form(func::VectorScalarAffineFunction{T}) where {T}
+    return convert(MOI.ScalarAffineFunction{T}, func)
+end
+
+"""
+    struct MatrixScalarQuadraticFunction{T, VT, MT} <: MOI.AbstractScalarFunction
```

Comment:
do we really need the full struct in the docstring?

Reply:
Fixed in latest commit